src/rulesets/LinearAlgebra/dense.jl

```diff
@@ -282,22 +282,349 @@ end
 ##### `norm`
 #####
 
-function rrule(::typeof(norm), A::AbstractArray{<:Real}, p::Real=2)
-    y = norm(A, p)
-    function norm_pullback(ȳ)
-        u = y^(1-p)
-        ∂A = @thunk ȳ .* u .* abs.(A).^p ./ A
-        ∂p = @thunk ȳ * (u * sum(a->abs(a)^p * log(abs(a)), A) - y * log(y)) / p
-        (NO_FIELDS, ∂A, ∂p)
+function frule((_, Δx, Δp), ::typeof(norm), x, p::Real)
+    return if isempty(x)
+        z = float(norm(zero(eltype(x))))
+        z, zero(z * zero(Δp) + z * zero(Δx))
+    elseif p == 2
```

Comment:
Arguably might want to use `===` here. If we are willing to commit to that require
The compiler has historically been better about constant folding  that (but i haven't checked recently).
and it is immune to invalidations i guess.

Reply:
Hm, I can't see a difference in benchmarks, but I'm not certain the best way to check on constant folding or invalidations. The arguments for using `==` that I can see are 1) it mirrors the primal function and 2) it enables the more efficient methods to be called even when the user calls `norm(x, 2.0)` instead of `norm(x, 2)` .